[5.2] Feat: Introduce Mail Template Layout

LadySolveig opened the pull request:
Pull Request for Issue #39492  .

### Summary of Changes

~~--The functionality for displaying the logo and sending it as an inline attachment depends on this fix #43828--~~ merged
~~--Depends on #43842 --~~ merged

Add support for mail template layouts as a parent wrapper around the existing mail templates.

- ✅  A new default responsive mail template **layout** file `mailtemplate.php` for rendering mail content with sitename in footer and optional logo.
- ✅  A new form field `MailtemplateLayoutField` to list available mail template layouts (overrides) from (frontend) templates.
- ✅  Updates to `MailTemplate` class to handle layout data and render email content using the specified layout.
- ✅   Configuration options in `config.xml` to enable/disable HTML layout and select a mail template layout globally.
- ✅   Form updates in `template.xml` to include options for disabling HTML layout and selecting a mail template layout individual for each template.
- ✅ Adjustments in `TemplateModel` to remove HTML layout fields when the mail style is set to plaintext.

### Testing Instructions

#### Global Configuration -> Mail Templates

Set `Mail Format`--> HTML

Setup global Mail Templates configration for Mail Template Layouts as follows:

![grafik](https://github.com/user-attachments/assets/7fae2906-2989-41de-9473-aec4d655bb02)

### Scenario 1

Per Template Settings are `disabled` in the Global Configuration:

![grafik](https://github.com/user-attachments/assets/8d229454-7e23-435b-8cd3-46b79a309ed6)

Go to `System`--> `Mail Templates`--> open `Global Configuration: Test Mail`
No additonal options in template edit view and the global settings are used.

![grafik](https://github.com/user-attachments/assets/0b2d7ba2-dd88-44e4-bf8d-ec726e712018)

Go to `System` --> `Global Configuration`--> `Tab: Server`

Scroll to the bottom and send out test mail

![grafik](https://github.com/user-attachments/assets/0f92ff54-628f-4ee9-8708-bf67a6fcb201)

Your Test Mail should look similar like this

![grafik](https://github.com/user-attachments/assets/2d5c2400-7a34-4a17-8e5b-7355b6cddc87)

### Scenario 2

Per Template Settings are `enabled` in the Global Configuration:
- additonal options in template edit view with default settings to use the global options

![grafik](https://github.com/user-attachments/assets/5332b25d-1ed2-42f0-bb2a-361c1d44c237)

![grafik](https://github.com/user-attachments/assets/30fcfed4-8a9d-4631-bd96-a7bfbc0c64f6)

Send the test mail again - the mail should look the same as in the previous step.

### Scenario 3

Go to `System`--> `Mail Templates`--> open `Global Configuration: Test Mail `

![grafik](https://github.com/user-attachments/assets/c219ba56-253d-44e4-85f7-01adc4b1e8dc)

Your Test Mail should look similar like this

![grafik](https://github.com/user-attachments/assets/30e778d7-ee46-4e0a-b687-29692b56777c)

### Scenario 4

Go to `System`--> `Site Templates`--> open your Template --> `Tab: Create Overrides` and add override `jooml->mail`

Open file html --> joomla --> mail --> mailtemplate.php

and add this text

![grafik](https://github.com/user-attachments/assets/da8067fc-3001-4958-9acf-70a776e04e1e)

`This is a test for template overrides`

Go to `System`--> `Mail Templates`--> open `Global Configuration: Test Mail`

Set Logo File --> `enables`

Choose your layout override

![grafik](https://github.com/user-attachments/assets/8be70dd2-dee1-45d9-8f97-7a2263ffe012)

Your Test Mail should look similar like this

![grafik](https://github.com/user-attachments/assets/5cf6ab3f-b80c-4e05-b807-74ed959755e0)

### Scenario 5

Go to `System`--> `Mail Templates`--> open `Global Configuration: Test Mail`

Choose Layout --> `Use Global`

Go to `System`--> `Global Configuration`--> `Mail Templates` 

Choose your layout override

Your Test Mail should look like in the step before

### Scenario 6

Go to `System`--> `Global Configuration`--> `Mail Templates` 

![grafik](https://github.com/user-attachments/assets/5212742a-43ce-4472-9abf-b860bf0997f2)

Your Test Mail should look similar like this

![grafik](https://github.com/user-attachments/assets/98695491-0f2a-49c9-a5f5-f0d94d67274a)

### Scenario 7

Go to `System`--> `Global Configuration`--> `Mail Templates`

Choose your layout override

![grafik](https://github.com/user-attachments/assets/04a75897-95a7-4707-b498-c7477e504249)

Go to `System`--> `Site Templates`--> open your Template

Open file html --> joomla --> mail --> mailtemplate.php

and add this text

![grafik](https://github.com/user-attachments/assets/12c53774-b21f-4bc6-8637-0b737730c567)

`{METHOD}`

Your Test Mail should look similar like this

![grafik](https://github.com/user-attachments/assets/c94ca984-e627-4f4d-9270-ddcf185bf958)


### Actual result BEFORE applying this Pull Request

![grafik](https://github.com/user-attachments/assets/98695491-0f2a-49c9-a5f5-f0d94d67274a)

### Expected result AFTER applying this Pull Request

![grafik](https://github.com/user-attachments/assets/2d5c2400-7a34-4a17-8e5b-7355b6cddc87)

### Link to documentations
Please select:
- [X] Documentation link for docs.joomla.org: <work in progress>
- [ ] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [ ] No documentation changes for manual.joomla.org needed

LadySolveig committed:
[5.2] [WIP] Feat: Add mail template layout as wrapper for mails

LadySolveig committed:
cs

LadySolveig committed:
fix cs

LadySolveig committed:
cs

brianteeman commented on administrator/components/com_mails/config.xml:
```diff
@@ -49,6 +50,37 @@
+			showon="mail_style:html[OR]mail_style:both[AND]disable_htmllayout:1"
+		/>
+
+		<field
+			name="mail_logoFile"
+			type="media"
+			schemes="http,https,ftp,ftps,data,file"
+			validate="url"
+			relative="true"
+			default=""
+			label="COM_CONFIG_FIELD_MAILTEMPLATE_LOGOFILE_LABEL"
+			showon="mail_style:html[OR]mail_style:both[AND]disable_htmllayout:1"
```
```suggestion
		<field
			name="mail_logoFile"
			type="media"
			label="COM_CONFIG_FIELD_MAILTEMPLATE_LOGOFILE_LABEL"
			schemes="http,https,ftp,ftps,data,file"
			validate="url"
			relative="true"
			default=""
			showon="mail_style:html[OR]mail_style:both[AND]disable_htmllayout:1"
```

LadySolveig replied:
Thanks!

LadySolveig committed:
Fix order in xml field

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented:
Can I suggest changing this
![image](https://github.com/user-attachments/assets/9113d1ba-08b8-4600-9c30-cd3c20a79158)

to
![image](https://github.com/user-attachments/assets/69412038-f3e2-4d15-8dbb-c63876a8dfbb)

LadySolveig commented:
@brianteeman if you want to change it by review in the language files I can commit directly :)

brianteeman commented on layouts/joomla/mail/mailtemplate.php:
```diff
@@ -0,0 +1,113 @@
+                            </tr>
+                            <tr>
+                                <td style="padding:30px;text-align:center;font-size:12px;background-color:#112855;color:#cccccc;">
+                                    <p style="margin:0;font-size:14px;line-height:20px;">&reg; <?php echo isset($extraData['siteName']) ? $extraData['siteName'] . ' ' : ''; ?><?php echo date("Y"); ?>
```
```suggestion
                                    <p style="margin:0;font-size:14px;line-height:20px;">&copy; <?php echo isset($extraData['siteName']) ? $extraData['siteName'] . ' ' : ''; ?><?php echo date("Y"); ?>
```

LadySolveig committed:
Update layouts/joomla/mail/mailtemplate.php

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented on administrator/language/en-GB/com_config.ini:
```diff
@@ -131,6 +131,9 @@ COM_CONFIG_FIELD_MAIL_SMTP_PASSWORD_LABEL="SMTP Password"
 COM_CONFIG_FIELD_MAIL_SMTP_PORT_LABEL="SMTP Port"
 COM_CONFIG_FIELD_MAIL_SMTP_SECURE_LABEL="SMTP Security"
 COM_CONFIG_FIELD_MAIL_SMTP_USERNAME_LABEL="SMTP Username"
+COM_CONFIG_FIELD_MAILTEMPLATE_LAYOUT_LABEL="Mail Template Layout"
```
```suggestion
COM_CONFIG_FIELD_MAILTEMPLATE_LAYOUT_LABEL="Layout"
```

LadySolveig replied:
Thank you 👍🏼

brianteeman commented on administrator/language/en-GB/com_config.ini:
```diff
@@ -131,6 +131,9 @@ COM_CONFIG_FIELD_MAIL_SMTP_PASSWORD_LABEL="SMTP Password"
 COM_CONFIG_FIELD_MAIL_SMTP_SECURE_LABEL="SMTP Security"
 COM_CONFIG_FIELD_MAIL_SMTP_USERNAME_LABEL="SMTP Username"
+COM_CONFIG_FIELD_MAILTEMPLATE_LAYOUT_LABEL="Mail Template Layout"
+COM_CONFIG_FIELD_MAILTEMPLATE_LAYOUT_OFF_LABEL="Disable Mail Template Layout"
```
```suggestion
COM_CONFIG_FIELD_MAILTEMPLATE_LAYOUT_OFF_LABEL="Mail Template Layout"
```

brianteeman commented on administrator/language/en-GB/com_mails.ini:
```diff
@@ -18,6 +18,8 @@ COM_MAILS_FIELD_BODY_LABEL="Body"
 COM_MAILS_FIELD_FILE_LABEL="File"
 COM_MAILS_FIELD_FILENAME_LABEL="File Name"
 COM_MAILS_FIELD_HTMLBODY_LABEL="HTML Body"
+COM_MAILS_FIELD_HTML_LAYOUT_LABEL="Mail Template Layout"
```
```suggestion
COM_MAILS_FIELD_HTML_LAYOUT_LABEL="Layout"
```

brianteeman commented on administrator/language/en-GB/com_mails.ini:
```diff
@@ -18,6 +18,8 @@ COM_MAILS_FIELD_BODY_LABEL="Body"
 COM_MAILS_FIELD_FILENAME_LABEL="File Name"
 COM_MAILS_FIELD_HTMLBODY_LABEL="HTML Body"
+COM_MAILS_FIELD_HTML_LAYOUT_LABEL="Mail Template Layout"
+COM_MAILS_FIELD_HTML_LAYOUT_OFF_LABEL="Disable Mail Template Layout"
```
```suggestion
COM_MAILS_FIELD_HTML_LAYOUT_OFF_LABEL="Mail Template Layout"
```

LadySolveig committed:
Update administrator/language/en-GB/com_config.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

LadySolveig committed:
Update administrator/language/en-GB/com_config.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

LadySolveig committed:
Update administrator/language/en-GB/com_mails.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

LadySolveig committed:
Update administrator/language/en-GB/com_mails.ini

Co-authored-by: Brian Teeman <brian@teeman.net>

brianteeman commented on layouts/joomla/mail/mailtemplate.php:
```diff
@@ -0,0 +1,113 @@
+        </style>
+    </head>
+    <body style="margin:0;padding:0;word-spacing:normal;background-color:#00000008;height:100%;">
+        <div role="article" aria-roledescription="email" lang="en" style="text-size-adjust:100%;-webkit-text-size-adjust:100%;-ms-text-size-adjust:100%;background-color:#00000008;height:100%;">
```
this should be using the lang of the email and not lang="en"

LadySolveig replied:
Nice catch - you are absolutely right. This should be dynamic. I create a fix for this.

brianteeman replied:
as you have set the lang on the parent html element you can remove it from here - unless the daft email formatting conditions require it here as well in which case please make it dynamic

LadySolveig replied:
Thank you!

brianteeman commented:
I am really happy to see this. It's the missing link to truly make branded html mails.

Basic functionality all works great.

brianteeman commented on layouts/joomla/mail/mailtemplate.php:
```diff
@@ -0,0 +1,113 @@
+                            <tr>
+                                <td style="padding:30px;text-align:center;font-size:12px;background-color:#112855;color:#cccccc;">
+                                    <p style="margin:0;font-size:14px;line-height:20px;">&copy; <?php echo isset($extraData['siteName']) ? $extraData['siteName'] . ' ' : ''; ?><?php echo date("Y"); ?>
+                                    <br><a title="<?php echo $siteUrl;?>" href="<?php echo $siteUrl; ?>" style="color:#cccccc;text-decoration:underline;"><?php echo substr_replace($siteUrl, 'www.', 0, strlen(Uri::getInstance()->toString(['scheme'])));?></a>
```
What is the purpose of enforcing www on the presentation of the site url. Not all sites use www especially subdomains

HLeithner replied:
it's the other way around, reducing the length of the visible link by removing the www. if it exists. Like it I would have done it easier including the schema but also ok (remove https://www. from https://google.com)

brianteeman replied:
That might have been the intent but its not what happens

![image](https://github.com/user-attachments/assets/a3d06771-590f-4e82-8ee7-60ec471ea631)

The substr_replace($siteUrl, 'www.', 0, strlen(Uri::getInstance()->toString(['scheme']))) function replaces the first occurrence of the URL's scheme (i.e., "http://" or "https://") with "www."

LadySolveig replied:
I'll be honest it just looked nicer than http:// 😄 , but of course you're right, technically it's just wrong. And I should put my personal preference in my own override rather than in the layout here. Changed 👍🏼

brianteeman commented on layouts/joomla/mail/mailtemplate.php:
```diff
@@ -0,0 +1,113 @@
+        </style>
+    </head>
+    <body style="margin:0;padding:0;word-spacing:normal;background-color:#00000008;height:100%;">
+        <div role="article" aria-roledescription="email" lang="en" style="text-size-adjust:100%;-webkit-text-size-adjust:100%;-ms-text-size-adjust:100%;background-color:#00000008;height:100%;">
```
~aria-roledescription="email"~

~Where did this come from - as far as I can see there is no aria role of email (https://www.w3.org/TR/wai-aria-1.1/#role_definitions) as a result this is invalid markup~

ignore that comment - i misunderstood how the roledescription is applied to the role

LadySolveig replied:
> aria-roledescription="email"

I think that's a misunderstanding. The aria role is `article`, only the aria-roledescription is email.

https://www.digitala11y.com/aria-roledescriptionproperties/

But I know that you are very well informed in this area. Would you want to remove it?
What do you think @chmst  - I would also be interested in your opinion on this?

brianteeman commented:
When per template email setting is set to NO then I would not expect to see the layout options per email - is that an error/bug

When per template email setting is set to YES I have the option to enable the layout and select a layout but I cannot select an image - is that an error/bug or by design. I am hoping its an error/bug as I would use log as an image header

brianteeman commented on layouts/joomla/mail/mailtemplate.php:
```diff
@@ -0,0 +1,113 @@
+
+/**
+ * @package     Joomla.Site
+ * @subpackage  com_mails
```
```suggestion
 * @subpackage  Layout
```

that's how all the other layouts are documented

Fedik commented:
Tanks @LadySolveig . That it is very good feature.

Some thoughts for future.
What also would be cool, is to have an API to add inline images more freely, before layout rendering (or within the layout).
Kind of:
```php
$mailTemplate->embedImages([
  'header-image' => 'path/to/image.jpg', 
  'logo-image' => 'path/to/logo.jpg'
]);
```
```html
// Then in layout:
<img src="cid:header-image" alt="inline image"/>
<img src="cid:logo-image" alt="logo image"/>
```
But I not sure how to correctly interconnect it.

Do not code it here, just ideas for future.  :wink:

LadySolveig committed:
Remove modification in layout that was replacing the scheme with  'www.' from the site URL link text

LadySolveig committed:
Update layouts/joomla/mail/mailtemplate.php

Co-authored-by: Brian Teeman <brian@teeman.net>

Fedik commented on libraries/src/Mail/MailTemplate.php:
```diff
@@ -274,6 +300,38 @@ public function send()
+                    $logo = Path::clean(JPATH_ROOT . '/' . htmlspecialchars(HTMLHelper::cleanImageURL($logo)->url));
+                    if (is_file($logo)) {
+                        # Attach the logo as inline attachement
+                        $this->mailer->addAttachment($logo, basename($logo), 'base64', mime_content_type($logo), 'inline');
```
Can also use dedicated method for this: `$this->mailer->addEmbeddedImage($path, $cid)`

LadySolveig replied:
Very nice 💚 I put it on my list and make a new PR

Fedik replied:
And I think, for the logo we better use fixed `cid`, kind of `site-logo`, 
it will be safer and esier than use of file name.

zero-24 commented on administrator/components/com_mails/config.xml:
```diff
@@ -4,7 +4,8 @@
 		name="mails_options"
-		label="COM_MAILS_CONFIG_MAIL_OPTIONS" >
+		label="COM_MAILS_CONFIG_MAIL_OPTIONS"
+		addfieldprefix="Joomla\Component\Mails\Administrator\Field" >
```
```suggestion
		addfieldprefix="Joomla\Component\Mails\Administrator\Field">
```

zero-24 commented on administrator/components/com_mails/src/Field/MailtemplateLayoutField.php:
```diff
@@ -0,0 +1,124 @@
+     */
+    protected function getInput()
+    {
+        // $extension = explode('.', $this->form->getValue('context'));
```
Is there a need to keep commented code

zero-24 commented on layouts/joomla/mail/mailtemplate.php:
```diff
@@ -0,0 +1,113 @@
+        <div role="article" aria-roledescription="email" lang="en" style="text-size-adjust:100%;-webkit-text-size-adjust:100%;-ms-text-size-adjust:100%;background-color:#00000008;height:100%;">
+            <table role="presentation" style="width:100%;border:none;border-spacing:0;height:100%;">
+                <tr>
+                    <td align="center" style="vertical-align:baseline; padding:30px 0">
```
I suspect its the outlook thing here but are that inline styles here in this file actually required?

Fedik replied:
Unfortunately, in many case it is only way to make something working in email :wink:

LadySolveig replied:
To my knowledge Gmail still strips the style tags. So it is safer to have it inline in Mails.

https://www.emailonacid.com/blog/article/email-development/12_things_you_must_know_when_developing_for_gmail_and_gmail_mobile_apps-2/

zero-24 commented on layouts/joomla/mail/mailtemplate.php:
```diff
@@ -0,0 +1,113 @@
+                        <tr>
+                        <td>
+                        <![endif]-->
+                        <table role="presentation" style="width:94%;max-width:630px;border:none;border-spacing:0;text-align:left;font-family:Arial,sans-serif;font-size:16px;line-height:22px;color:#363636;">
```
Fore example font family is quite selective shouldt we let that be down to the client who displays the item? Similiar with width: 94% or max-width etc etc.

LadySolveig replied:
If you remove the width, it is not responsive anymore. At the moment, 600px is the standard width that is currently considered best practice in email marketing.

LadySolveig replied:
If I remove the default font (Websafe - Arial) here, it will be displayed differently on each client.
Then we are again far away from a layout that is uniformly displayed on every device.
That would be similar to delivering Cassiopeia without a default font.

I completely agree with you that this is of course very individual, but there is the option of overriding in your template. 

I would like to offer a good standard for those who don't want to do it themselves and perhaps have no idea of the differences and pitfalls when creating an HTML email.

Fedik replied:
@zero-24 It is common practice in the world of the email templating.
Some devs goes much further and doing it for evry `<td>`.

Email clients is diferent from real Browsers  very much.
We cannot escape from inline styles, and hacky css tricks.

LadySolveig replied:
I don't like it either, but unfortunately it is necessary.

zero-24 replied:
Ok thanks. Didnt knew all the details only that it is a mess. Thanks for checking up on that. 👍

zero-24 commented on libraries/src/Mail/MailTemplate.php:
```diff
@@ -245,10 +270,11 @@ public function send()
         $plainData = $this->plain_data ?: $this->data;
         $plainBody = $this->replaceTags(Text::_($mail->body), $plainData);
-        $htmlBody  = $this->replaceTags(Text::_($mail->htmlbody), $this->data);
+        $htmlBody  = $useLayout ? Text::_($mail->htmlbody) : $this->replaceTags(Text::_($mail->htmlbody), $this->data);
```
On a first look it seams that the replacement of variables would not work anymore with that change? Or whats the need for the change here?

LadySolveig replied:
That you can also use specific tags in your overrides and they will be replaced after rendering the layout. Useful if you have a specific template override for an email.
But otherwise it doesn't hurt too.
The replacement in this case is in line 332

zero-24 commented on libraries/src/Mail/MailTemplate.php:
```diff
@@ -274,6 +300,38 @@ public function send()
+                }
+
+                $layout = $params->get('htmllayout', $config->get('mail_htmllayout', 'mailtemplate'));
+                // Wrap the default Joomla mail template around the HTML body
```
```suggestion

                // Wrap the default Joomla mail template around the HTML body
```

zero-24 commented on libraries/src/Mail/MailTemplate.php:
```diff
@@ -274,6 +300,38 @@ public function send()
+
+                $layout = $params->get('htmllayout', $config->get('mail_htmllayout', 'mailtemplate'));
+                // Wrap the default Joomla mail template around the HTML body
+                $layoutFile   = new FileLayout('joomla.mail.' . $layout, null, ['client' => 'site']);
```
```suggestion
                $layoutFile  = new FileLayout('joomla.mail.' . $layout, null, ['client' => 'site']);
```

zero-24 commented on libraries/src/Mail/MailTemplate.php:
```diff
@@ -274,6 +300,38 @@ public function send()
+                $layout = $params->get('htmllayout', $config->get('mail_htmllayout', 'mailtemplate'));
+                // Wrap the default Joomla mail template around the HTML body
+                $layoutFile   = new FileLayout('joomla.mail.' . $layout, null, ['client' => 'site']);
+                // Set the default template search path
```
```suggestion

                // Set the default template search path
```

zero-24 commented on libraries/src/Mail/MailTemplate.php:
```diff
@@ -274,6 +300,38 @@ public function send()
+
+                $htmlBody = $layoutFile->render(['mail' => $htmlBody, 'extra' => $this->layoutTemplateData], null);
+
+                $htmlBody  = $this->replaceTags(Text::_($htmlBody), $this->data);
```
```suggestion
                $htmlBody = $this->replaceTags(Text::_($htmlBody), $this->data);
```

coolcat-creations commented:
Awesome, can one create a signature somewhere and embed it in the mail template?
Thank you!!!

brianteeman commented:
> Awesome, can one create a signature somewhere and embed it in the mail template? Thank you!!!

You would need to create your own mail template for that

coolcat-creations commented:
Maybe this: ® Joomla mail 2024
www.localhost/mail/ is a language string and could be used?
I will check it out as soon as possible.

brianteeman commented:
To answer comments about the html and inline styles used. @LadySolveig is absolutely correct html in email is stuck back in the 90's. I ran the generated email through a checker (mailtrap.io) 

![image](https://github.com/user-attachments/assets/78ab8b0b-9c4f-4808-8c6d-4f542b1f3d31)

LadySolveig commented:
Thank you for the additional check @brianteeman 
I will address your other comment later this evening and give an update 👍🏼

brianteeman commented:
> Maybe this: ® Joomla mail 2024 [www.localhost/mail/](http://www.localhost/mail/) is a language string and could be used? I will check it out as soon as possible.

it is not a string. As already stated you would need to create a layout

LadySolveig commented:
@coolcat-creations you can create a complete layout override in your frontend template for this purpose 👍🏼

coolcat-creations commented:
love it!

Hackwar committed:
Merge branch '5.2-dev' into 5.2/feat/mailtemplate

webnet-assmann commented:
I have tested this item :white_check_mark: successfully on abe2893fbcdc23f42f15d70cb3cb15f5ec4c1aea<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43829">issues.joomla.org/tracker/joomla-cms/43829</a>.</sub>

sayfrndship1234 commented:
great effort !

LadySolveig committed:
Merge remote-tracking branch 'origin/5.2-dev' into 5.2/feat/mailtemplate

LadySolveig committed:
various improvements, fixes and additional option `disable logofile`

- Add additional option to disable attachement of logofile to mail and useglobal in template.xml
- point to the right template of the selected layout override
- only show additional options per Mail Template if option is diabled in Global Configuration

LadySolveig committed:
phpcs-fixer + suggestions from review

LadySolveig committed:
ISO 2 Letter Language Code dynamic in layout

LadySolveig committed:
Fixed cid 'site-logo' for attached logo image

LadySolveig committed:
Allow image files with whitespace and rename variable name

LadySolveig committed:
Merge remote-tracking branch 'myjoomla/5.2/feat/mailtemplate' into 5.2/feat/mailtemplate

LadySolveig commented:
@brianteeman I am looking forward to your feedback :)

I have now reworked and hopefully incorporated the use case you described as practicably as possible.
(Short note: My changes are partly dependent on this PR #43842, so that only the fields that are relevant are displayed - whenever possible - and there is as little confusion as possible due to the additional options.

## Scenario 1

Per Template Settings are `disabled` in the Global Configuration:
- no additonal options in template edit view and the global settings are used.

![grafik](https://github.com/user-attachments/assets/8d229454-7e23-435b-8cd3-46b79a309ed6)

![grafik](https://github.com/user-attachments/assets/0b2d7ba2-dd88-44e4-bf8d-ec726e712018)

## Scenario 2

Per Template Settings are `enabled` in the Global Configuration:
- additonal options in template edit view with default settings to use the global options
- option to disable the sending of the logofile as an attachment. (If no logo is selected in the global configuration, it will not be sent anyway)
- separated from the `Mail Settings` switcher:

As I noticed during testing that PHP Mail is always entered for the Mailer option when it `Mail Settings` is activated (completely independent of the settings in the global configuration).
I don't find it particularly user-friendly if I “only” want to change the layout and then have to check the mailer settings also per template. I consider this to be a major potential source of errors in this context.
But I am open to objections and other views here.

![grafik](https://github.com/user-attachments/assets/5332b25d-1ed2-42f0-bb2a-361c1d44c237)

![grafik](https://github.com/user-attachments/assets/30fcfed4-8a9d-4631-bd96-a7bfbc0c64f6)

LadySolveig committed:
dynamic lang code part 2

LadySolveig committed:
cs and remove second lang definition from layout

Quy commented on layouts/joomla/mail/mailtemplate.php:
```diff
@@ -0,0 +1,113 @@
+        </style>
+    </head>
+    <body style="margin:0;padding:0;word-spacing:normal;background-color:#00000008;height:100%;">
+        <div role="article" aria-roledescription="email" lang="<?php echo (isset($extraData['lang'])) ?  $extraData['lang'] : 'en' ?>" style="text-size-adjust:100%;-webkit-text-size-adjust:100%;-ms-text-size-adjust:100%;background-color:#00000008;height:100%;">
```
```suggestion
        <div role="article" aria-roledescription="email" lang="<?php echo isset($extraData['lang']) ?  $extraData['lang'] : 'en'; ?>" style="text-size-adjust:100%;-webkit-text-size-adjust:100%;-ms-text-size-adjust:100%;background-color:#00000008;height:100%;">
```

Quy commented on layouts/joomla/mail/mailtemplate.php:
```diff
@@ -0,0 +1,113 @@
+        </style>
+    </head>
+    <body style="margin:0;padding:0;word-spacing:normal;background-color:#00000008;height:100%;">
+        <div role="article" aria-roledescription="email" lang="<?php echo (isset($extraData['lang'])) ?  $extraData['lang'] : 'en' ?>" style="text-size-adjust:100%;-webkit-text-size-adjust:100%;-ms-text-size-adjust:100%;background-color:#00000008;height:100%;">
```
```suggestion
        <div role="article" aria-roledescription="email" lang="<?php echo isset($extraData['lang']) ? $extraData['lang'] : 'en'; ?>" style="text-size-adjust:100%;-webkit-text-size-adjust:100%;-ms-text-size-adjust:100%;background-color:#00000008;height:100%;">
```

LadySolveig replied:
Thank you @Quy I decided to remove this completely cause it is not really necessary.

Fedik commented on libraries/src/Mail/MailTemplate.php:
```diff
@@ -274,6 +305,57 @@ public function send()
+
+                // Add the logo to the mail as inline attachement
+                if ($logo) {
+                    $logo = Path::check(JPATH_ROOT . '/' . HTMLHelper::cleanImageURL($logo)->url);
```
Please HtmlHelper as `HTMLHelper::_()`: 
```php
HTMLHelper::_('cleanImageURL', $logo)->url;
```

LadySolveig committed:
Fix no tatic HtmlHelper

bembelimen commented:
I have tested this item :white_check_mark: successfully on 1e2cb33de6bd104cd7b813f848b2e67323d48b7d<br /><br />Also tested with child-template overrides. Works great. Thanks for the PR!<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43829">issues.joomla.org/tracker/joomla-cms/43829</a>.</sub>

LadySolveig commented:
If you have time, I can ask you to repeat your test again. I have revised the testing instuctions and the PR once again. @webnet-assmann Thank you 💚

brianteeman commented:
Why is it that when per template is enabled I can only chose the layout and enable/disable the logo. Why can I not change the logo

LadySolveig commented:
I thought of a single place for the logo brand. and only for the deactivation so that it is not attached if the logo is not present in the template override.

Would prefer to create a follow up PR for this additional option in conjunction with @Fedik  suggestion for an Image API here https://github.com/joomla/joomla-cms/pull/43829#issuecomment-2245172520

brianteeman commented:
If I can change the mail template I should also be able to change the image file. For example the dimensions of the image file might not be appropriate for every template but there are many reasons to want to change the image used. Am I missing some reasoning that explains why I woudl not want to do that

LadySolveig commented:
It was a really good point that it is possible to have an override without an image, and then the image shouldn't be attached either.
As in your example, outputting the log in the header.
But changing the image per template is an improvement that is outside the scope of this PR.

As mentioned, that would be a good improvement for an additional PR, as also adding an API for the inline attachments.

brianteeman commented:
I dont understand your logic for not allowing the image to be changed. To me it devalues the usefulness of having per mail layouts but I am not going to block it

coolcat-creations commented:
how about solving the image "issue" like that, that the image must be always in a path with the template name? I totally see the need for different logos on several sites.

bembelimen commented:
Why then not just making a PR when this is merged?

coolcat-creations commented:
> Why then not just making a PR when this is merged?

That would be fine too — otherwise the feature is lost in discussion.

brianteeman commented:
I have tested this item :white_check_mark: successfully on 1e2cb33de6bd104cd7b813f848b2e67323d48b7d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43829">issues.joomla.org/tracker/joomla-cms/43829</a>.</sub>

chmst commented:
I have tested this item :white_check_mark: successfully on 1e2cb33de6bd104cd7b813f848b2e67323d48b7d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43829">issues.joomla.org/tracker/joomla-cms/43829</a>.</sub>

chmst commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43829">issues.joomla.org/tracker/joomla-cms/43829</a>.</sub>

exlemor commented:
I have tested this item :white_check_mark: successfully on 1e2cb33de6bd104cd7b813f848b2e67323d48b7d<br /><br />I was able to test this successfully.

As discussed with Martina, I should note that Scenarios 1-5, I had my Mailer set to PHP Mail and they worked fine, Scenario 6 gave an Could not instantiate mail function but when I switched the mailer to Sendmail that scenario worked. Scenario 7 worked on PHP Mail.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43829">issues.joomla.org/tracker/joomla-cms/43829</a>.</sub>

HLeithner commented on administrator/components/com_mails/src/Field/MailtemplateLayoutField.php:
```diff
@@ -0,0 +1,131 @@
+        $query = $db->getQuery(true);
+
+        // Build the query.
+        $query->select('element, name')
+            ->from('#__extensions')
+            ->where('client_id = 0')
+            ->where('type = ' . $db->quote('template'))
+            ->where('enabled = 1');
```
quoteName missing

LadySolveig replied:
Thanks - additional #43851

LadySolveig replied:
could you please resolve if it is ok for you now? or should I change something else?

HLeithner commented on libraries/src/Mail/MailTemplate.php:
```diff
@@ -274,6 +305,57 @@ public function send()
+                    $logo   = $params->get('disable_logofile', 1) ? $logo : '' ;
+                }
+
+                // Add the logo to the mail as inline attachement
+                if ($logo) {
+                    $logo = Path::check(JPATH_ROOT . '/' . HTMLHelper::_('cleanImageURL', $logo)->url);
+                    if (is_file(urldecode($logo))) {
+                        # Attach the logo as inline attachement
+                        $this->mailer->addAttachment($logo, 'site-logo', 'base64', mime_content_type($logo), 'inline');
+
+                        // We need only the cid for attached logo file
+                        $this->addLayoutTemplateData(['logo' => 'site-logo']);
+                    }
+                }
```
just as side note, phpmailer can automatically attach images as inline files from html body

Fedik replied:
That is interesting, that could help a loot. Have a link to some example?

HLeithner replied:
I only checked the phpmailer source code (I used this before so I know it's working) it only take relative urls

HLeithner commented on libraries/src/Mail/MailTemplate.php:
```diff
@@ -274,6 +305,57 @@ public function send()
+                    ]);
+                }
+
+                $htmlBody = $layoutFile->render(['mail' => $htmlBody, 'extra' => $this->layoutTemplateData], null);
```
I miss the $mailer object in the layout, i would allow adding own stuff like attachments and additional emails much easier.

Fedik replied:
I also thought about it. But I think it should not be allowed. 
Because it breaks whole concept of layout system.
The layout is the "end-point", just rendering, nothing more.

LadySolveig replied:
Is there a particular reason why you don't want to use the `onMailBeforeRendering` event for this? 
I don't see a problem here to add additional attachments and also additional data via addLayoutTemplateData. Or am I missing something here?
I don't know if I really want more logic in a layout than it needs.

HLeithner replied:
maybe it's not needed because an event exists, on the other side I use this in the template for setting the subject for example but that's a shop system that might be too complex situation.

HLeithner reviewed (changes requested):
question in general, why do we need a layout.php file and not use the mail template component it self and save the base html layout in the database so people can adapt it easily at the place where all other mail layouts are stored?

Fedik commented:
> save the base html layout in the database so people can adapt it easily 

No please, it is like saving index.php of the main template in database.
It will be huge and unpractical.

HLeithner commented:
> > save the base html layout in the database so people can adapt it easily
> 
> No please, it is like saving index.php of the main template in database. It will be huge and unpractical.

tbh, we need to be more user friendly for end users and they are unable (or don't want to) to touch php documents.
In the provided layout is nothing fancy that needs php.

Fedik commented:
This kind of thing is for "advanced user", it is not average HTML that can be edited in TinyMCE.
However the layout can be provided as part of the site template (by template developers), and it can be edited as any other template file.

HLeithner commented:
yeah don't care anyway please just fix the broken database query thanks

LadySolveig committed:
Fix db->quote missing

brianteeman commented:
I suspect most people will just be happy to have real html emails with the default layout - I know I am

LadySolveig committed:
remove empty space

HLeithner reviewed (approved):
thx

LadySolveig committed:
improve query

Hackwar committed:
Merge branch '5.2-dev' into 5.2/feat/mailtemplate

Hackwar commented:
I'm merging this even though since this was set to RTC, additional changes were made. However those changes were basically codestyle fixes and have been reviewed by me.

Thank you for your contribution @LadySolveig! This will be a great feature.

rbuelund commented:
I have tried the new main template feature, but I cannot seem to find the posibility for an override of the layout in the Cassiopeia template ?? Where should that be ?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/43829">issues.joomla.org/tracker/joomla-cms/43829</a>.</sub>

rbuelund commented:
Ahh sorry - found it now.
